### PR TITLE
Bundler ignores extra nonrelevant credentials

### LINKED ISSUE #2657: Dependabot fails to update "bundler" dependencies if the credentials array contains a Docker registry
**Package manager/ecosystem**
ruby:bundler

**Manifest contents prior to update**
```
ruby "2.6.5"
source "https://rubygems.org"

gem "nokogiri", "~> 1.9.0"
```

**Updated dependency**
Not applicable

**What you expected to see, versus what you actually saw**

Based on the [generic-update-script.rb](https://github.com/dependabot/dependabot-script/blob/main/generic-update-script.rb), I'm writing / maintaining my own script for updating dependencies hosted in my private Gitlab. In that script, I add several credentials for the different private repositories / servers that are used, including the credentials for accessing the Docker registry in Gitlab, based on the Dependabot test case in [dependabot-core/docker/spec/dependabotdocker/update_checker_spec.rb](https://github.com/dependabot/dependabot-core/blob/3d4d7d5ac7743a799a0951a53da4894ce59fc1ff/docker/spec/dependabot/docker/update_checker_spec.rb#L554):

```
credentials << {
    "type" => "docker_registry",
    "registry" => ENV["CI_REGISTRY"],
    "username" => "gitlab-ci-token",
    "password" => ENV["CI_JOB_TOKEN"] # The token with access permission for the Gitlab Docker registry
}
```

However, if I have these Docker related credentials defined when checking for updates for `bundler` dependencies, then Dependabot fails with the following stacktrace. This problem started to occur after upgrading from dependabot-core 122.1 to dependabot-core 123.0.

```
/home/dependabot/vendor/ruby/2.6.0/gems/dependabot-common-0.123.0/lib/dependabot/shared_helpers.rb:113:in `run_helper_subprocess': key not found: "host" (Dependabot::SharedHelpers::HelperSubprocessFailed)
	from /home/dependabot/vendor/ruby/2.6.0/gems/dependabot-bundler-0.123.0/lib/dependabot/bundler/update_checker/latest_version_finder/dependency_source.rb:121:in `block in source_type'
	from /home/dependabot/vendor/ruby/2.6.0/gems/dependabot-bundler-0.123.0/lib/dependabot/bundler/update_checker/shared_bundler_helpers.rb:53:in `block in in_a_native_bundler_context'
	from /home/dependabot/vendor/ruby/2.6.0/gems/dependabot-common-0.123.0/lib/dependabot/shared_helpers.rb:56:in `block (2 levels) in in_a_temporary_directory'
	from /home/dependabot/vendor/ruby/2.6.0/gems/dependabot-common-0.123.0/lib/dependabot/shared_helpers.rb:56:in `chdir'
	from /home/dependabot/vendor/ruby/2.6.0/gems/dependabot-common-0.123.0/lib/dependabot/shared_helpers.rb:56:in `block in in_a_temporary_directory'
	from /usr/lib/ruby/2.6.0/tmpdir.rb:93:in `mktmpdir'
	from /home/dependabot/vendor/ruby/2.6.0/gems/dependabot-common-0.123.0/lib/dependabot/shared_helpers.rb:53:in `in_a_temporary_directory'
	from /home/dependabot/vendor/ruby/2.6.0/gems/dependabot-common-0.123.0/lib/dependabot/shared_helpers.rb:47:in `in_a_temporary_repo_directory'
	from /home/dependabot/vendor/ruby/2.6.0/gems/dependabot-bundler-0.123.0/lib/dependabot/bundler/update_checker/shared_bundler_helpers.rb:49:in `in_a_native_bundler_context'
	from /home/dependabot/vendor/ruby/2.6.0/gems/dependabot-bundler-0.123.0/lib/dependabot/bundler/update_checker/latest_version_finder/dependency_source.rb:120:in `source_type'
	from /home/dependabot/vendor/ruby/2.6.0/gems/dependabot-bundler-0.123.0/lib/dependabot/bundler/update_checker/latest_version_finder/dependency_source.rb:73:in `git?'
	from /home/dependabot/vendor/ruby/2.6.0/gems/dependabot-bundler-0.123.0/lib/dependabot/bundler/update_checker/latest_version_finder.rb:42:in `fetch_latest_version_details'
	from /home/dependabot/vendor/ruby/2.6.0/gems/dependabot-bundler-0.123.0/lib/dependabot/bundler/update_checker/latest_version_finder.rb:29:in `latest_version_details'
	from /home/dependabot/vendor/ruby/2.6.0/gems/dependabot-bundler-0.123.0/lib/dependabot/bundler/update_checker.rb:185:in `latest_version_details'
	from /home/dependabot/vendor/ruby/2.6.0/gems/dependabot-bundler-0.123.0/lib/dependabot/bundler/update_checker.rb:20:in `latest_version'
	from /home/dependabot/vendor/ruby/2.6.0/gems/dependabot-common-0.123.0/lib/dependabot/update_checkers/base.rb:259:in `can_compare_requirements?'
	from /home/dependabot/vendor/ruby/2.6.0/gems/dependabot-common-0.123.0/lib/dependabot/update_checkers/base.rb:249:in `requirements_up_to_date?'
	from /home/dependabot/vendor/ruby/2.6.0/gems/dependabot-common-0.123.0/lib/dependabot/update_checkers/base.rb:32:in `up_to_date?'
```

I would expect that the code for updating `bundler` dependencies is gracefully ignoring all credentials that have an irrelevant type like `docker_registry`. Instead, Dependabot is currently failing when checking for updates of `bundler` dependencies if `docker_registry` credentials are defined. Reason is that `docker_registry` credentials are not containing a `host` key-value pair but instead use `registry` for identifying the host.

If I modify the above Docker related credential so that it also contains a `host` entry, then Dependabot is able to check for `bundler` dependencies again:

```
credentials << {
    "type" => "docker_registry",
    "registry" => ENV["CI_REGISTRY"],
    "host" => ENV["CI_REGISTRY"],
    "username" => "gitlab-ci-token",
    "password" => ENV["CI_JOB_TOKEN"] # The token with access permission for the Gitlab Docker registry
}
```

**Images of the diff or a link to the PR, issue or logs**

This problem has been introduced in dependabot-core v0.123.0 when the function `set_bundler_flags_and_credentials` has been introduced in [bundler/helpers/lib/functions.rb](https://github.com/dependabot/dependabot-core/blob/3d4d7d5ac7743a799a0951a53da4894ce59fc1ff/bundler/helpers/lib/functions.rb#L138):

```
...

credentials.each do |cred|
  token = cred["token"] ||
          "#{cred['username']}:#{cred['password']}"

  Bundler.settings.set_command_option(
    cred.fetch("host"),
    token.gsub("@", "%40F").gsub("?", "%3F")
  )
end

...
```

I would suggest to add a filter here so that only credentials with the relevant type are getting processed, i.e. I would assume only credentials of type `rubygems_server` should be processed here?

### PR DESCRIPTION
**Pull request by baseballlover723**
Fixes #2657. There probably ought to be another test file, or this test added in more places, to be more complete, but I'm can't spend any more time, so this will have to do for now.

### CODE REVIEW

**Comment by baseballlover723** [comment]
I don't know why tests for other languages keep failing / timing out. The bundler tests pass, which should be the only ones that matter imo.

**Comment by baseballlover723** [comment]
seems like there might be resource issue with the github actions runner. exit code 137 seems to be the usual code when a process is killed due to being out of memory. You may want to restart your github actions runners / see if theres anything wrong with them.

**Comment by jurre** [comment]
Yeah I don't think the test failures are introduced here, no worries!

It seems like we already do this filtering before we pass them off to the native helpers, [here](https://github.com/dependabot/dependabot-core/blob/main/bundler/lib/dependabot/bundler/file_updater/lockfile_updater.rb#L237), [here](https://github.com/dependabot/dependabot-core/blob/15dd03b79b59fc500e625c5c8a4136b273401abc/bundler/lib/dependabot/bundler/update_checker/force_updater.rb#L144), [here](https://github.com/dependabot/dependabot-core/blob/main/bundler/lib/dependabot/bundler/update_checker/conflicting_dependency_resolver.rb#L38) (via SharedBundlerHelpers) and [here](https://github.com/dependabot/dependabot-core/blob/main/bundler/lib/dependabot/bundler/update_checker/shared_bundler_helpers.rb#L213)

Definitely some duplication we may be able to get rid of, but I don't immediately see which extra entries this PR filters out?

**Comment by baseballlover723** [comment]
I'm not too sure exactly where, but somewhere there something missing, and some non relevant credentials are getting in. I think there was a file that didn't have the filtering? but I could be misremembering.

>  but I don't immediately see which extra entries this PR filters out?

I'm not too sure what you mean by this

**Comment by baseballlover723** [comment]
What are the next steps for this PR?

**Comment by jurre** [comment]
> What are the next steps for this PR?

I'm not 100% clear on what problem this PR is solving, could you maybe describe the scenario in which this is an issue? When exactly do we pass credentials to bundler that we shouldn't?

**Comment by feelepxyz** [comment]
> I'm not 100% clear on what problem this PR is solving, could you maybe describe the scenario in which this is an issue? When exactly do we pass credentials to bundler that we shouldn't?

I believe this is for self-hosted versions of core where we pass through all credentials. Seems like a good change to only filter in native helper functions.

It looks like we already do this in a few places that could probably be cleaned up with this change.

We should be able to pass in `credentials` directly to these methods and delete `relevant_credentials`:
- https://github.com/dependabot/dependabot-core/blob/b7c25bc8c776d6571b63b645a89871f5d0b8501b/bundler%2Flib%2Fdependabot%2Fbundler%2Ffile_updater%2Flockfile_updater.rb#L75
- https://github.com/dependabot/dependabot-core/blob/3197f2f36eb56e993fe1777683ed2fdf5d61f2a8/bundler%2Flib%2Fdependabot%2Fbundler%2Fupdate_checker%2Fconflicting_dependency_resolver.rb#L38
- https://github.com/dependabot/dependabot-core/blob/c61c4948a86fb4196e211070a370148e028df76d/bundler%2Flib%2Fdependabot%2Fbundler%2Fupdate_checker%2Fforce_updater.rb#L53
- https://github.com/dependabot/dependabot-core/blob/b7c25bc8c776d6571b63b645a89871f5d0b8501b/bundler%2Flib%2Fdependabot%2Fbundler%2Fupdate_checker%2Fshared_bundler_helpers.rb#L172
- https://github.com/dependabot/dependabot-core/blob/b7c25bc8c776d6571b63b645a89871f5d0b8501b/bundler%2Flib%2Fdependabot%2Fbundler%2Fupdate_checker%2Fshared_bundler_helpers.rb#L196

**Comment by feelepxyz** [comment]
> I'm not too sure exactly where, but somewhere there something missing, and some non relevant credentials are getting in. I think there was a file that didn't have the filtering?

Looks like this is happening here:
- https://github.com/dependabot/dependabot-core/blob/b7c25bc8c776d6571b63b645a89871f5d0b8501b/bundler%2Flib%2Fdependabot%2Fbundler%2Fupdate_checker%2Fversion_resolver.rb#L88
- https://github.com/dependabot/dependabot-core/blob/7d08fea5fabf31b3b959c356db459af0f66739e6/bundler%2Flib%2Fdependabot%2Fbundler%2Fupdate_checker%2Flatest_version_finder%2Fdependency_source.rb#L63
- https://github.com/dependabot/dependabot-core/blob/7d08fea5fabf31b3b959c356db459af0f66739e6/bundler%2Flib%2Fdependabot%2Fbundler%2Fupdate_checker%2Flatest_version_finder%2Fdependency_source.rb#L108
- https://github.com/dependabot/dependabot-core/blob/7d08fea5fabf31b3b959c356db459af0f66739e6/bundler%2Flib%2Fdependabot%2Fbundler%2Fupdate_checker%2Flatest_version_finder%2Fdependency_source.rb#L128

**Comment by baseballlover723** [comment]
yes as @feelepxyz we are using dependabot in a self hosted instance, and we set our credentials for all of our registries into the same env variable, which is used for all of the dependabot jobs. As noted by @feelepxyz, there are some places where filtering wasn't happening. IMO its better to do the filtering in the shared helpers, since then you don't have to worry about possibility missing a place where its filtered, since it should only consider relevant credentials always. I'll remove the other filtering and just leave the shared helper filtering.

**Comment by baseballlover723** [comment]
@jurre this is ready to be reviewed when you get a chance.

File: bundler/helpers/lib/functions.rb [hunk 1]
```diff
@@ -163,6 +163,24 @@ def self.set_bundler_flags_and_credentials(dir:, credentials:,
...
+  end
+
+  def self.private_registry_credentials(credentials)
+    credentials.
+      select { |cred| cred["type"] == "rubygems_server" }
+  end
+
+  def self.git_source_credentials(credentials)
+    credentials.
+      select { |cred| cred["password"] || cred["token"] }.
```
Comment (feelepxyz) [inline bundler/helpers/lib/functions.rb:180]: Minor but looks like this is already done for both under `relevant_credentials` so could ✂️ 
```suggestion
```

**Review by feelepxyz — APPROVED** [review]
This looks good to me 👍  Thanks for fixing this and taking on the feedback 🙇

**Comment by baseballlover723** [comment]
not a problem, glad to help out

**Comment by feelepxyz** [comment]
@baseballlover723 could you rebase/merge in the latest changes from `main`?

**Comment by baseballlover723** [comment]
@feelepxyz done

**Comment by feelepxyz** [comment]
> @feelepxyz done

@baseballlover723 sorry CI has been breaking, I think it's fixed on main now, mind rebasing again?

**Comment by baseballlover723** [comment]
@feelepxyz done. Glad to see CI fixed

**Pull request merged**